### PR TITLE
Allow Spring Boot to determine type created by GatewayProxyFactoryBean

### PR DESCRIPTION
**Pull request by wilkinsona**
This PR fixes the Spring Integration portion of spring-projects/spring-boot#2811. Please see the commit message for further details. I've tested the changes against a Spring Boot 1.2.4 snapshot and confirmed that it resolves the problem.

### CODE REVIEW

**Comment by garyrussell** [comment]
https://jira.spring.io/browse/INT-3701

File: spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java [hunk 1]
```diff
@@ -163,7 +164,10 @@ else if (StringUtils.hasText(asyncExecutor)) {
 
 		gatewayProxyBuilder.addConstructorArgValue(serviceInterface);
 
-		return new BeanDefinitionHolder(gatewayProxyBuilder.getBeanDefinition(), id);
+		AbstractBeanDefinition beanDefinition = gatewayProxyBuilder.getBeanDefinition();
+		beanDefinition.addMetadataAttribute(new BeanMetadataAttribute("factoryBeanObjectType", serviceInterface));
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java]: Perhaps we should open issues against Spring Framework and Spring Boot to move the constant `FACTORY_BEAN_OBJECT_TYPE` to SF - e.g. to `BeanMetadataAttribute`.

It's a bit brittle to use a literal here and it is possible that other projects may need similar.

For now, I will move this to a constant within SI during the merge.

I will also address @artembilan 's suggestion about fixing the `getObjectType()` when the service interface is null.

Comment (wilkinsona) [inline spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java]: On the Boot call yesterday, @philwebb mentioned some of the code moving into Spring Framework. It hasn't happened yet, AFAIK. Hopefully Phil can comment with the latest.

Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java]: Note that this solution won't work if the service interface is a property placeholder (rare, perhaps, but possible).

We may just have to live with that, or boot will have to resolve property placeholders in `BeanTypeRegistry.determineTypeFromDefinitionAttribute()`.

**Comment by garyrussell** [comment]
With polish: https://github.com/garyrussell/spring-integration/commit/29a6bf544bf595a15ec49e8e90fea0a07b7b912e

/cc @artembilan

**Comment by artembilan** [comment]
Guys, nothing to complain about.
If @wilkinsona confirms, that the fix work well with Boot, we can go ahead and merge.

@garyrussell , be careful with back port to `4.1.x`: your simple change for the `reactor.fn.Functions` import order may course conflict. `4.1` is based on the Reactor `1.0`, where `Functions` is on the other package.
From other side: do we really need to backport this?
@wilkinsona , do we need a fix for Boot `1.2.x` ?

Thank you both

**Comment by garyrussell** [comment]
@artembilan -   @wilkinsona 's PR is against 4.1.x so it certainly sounds like it's needed in 4.1

I just did my polishing on master hence, the reactor issue - will be sure to verify on backport

**Comment by artembilan** [comment]
Doh. I have missed that from the PR header.
Thank you and I believe you. 
So, please, go ahead and merge

**Comment by garyrussell** [comment]
Reverting the change to the `GPFB` - it caused an undesired side-effect (early access to the type should return null).

**Comment by garyrussell** [comment]
Merged as 2653ce9aed1f19c52fed349cefc5a7315cbe29e2 and cherry-picked to 4.1.x.

cc: spring-projects/spring-boot#2811

**Comment by chrylis** [comment]
Much appreciated! What will be the minimum release versions required for this to work?

**Comment by garyrussell** [comment]
4.1.4; we don't currently have a schedule for its release (we only just released 4.1.3). 

When do you need it?

In the meantime, you can test with 4.1.4.BUILD-SNAPSHOT (from the snapshot repo).

**Comment by garyrussell** [comment]
Just to be clear, that's Spring Integration - we'll need to coordinate with the boot guys for a boot version.

File: spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java (hunk 1 above)
Comment (philwebb) [inline spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java]: @wilkinsona I was thinking of https://jira.spring.io/browse/SPR-11480

**Comment by chrylis** [comment]
I have a library to update when it's ready. I'm keeping the old-style explicit import for now instead of using autoconfiguration.

**Pull request closed without merging**